openedx/adg/lms/applications/views.py

```diff
@@ -1,3 +1,174 @@
+
+        """
+        if not request.user.is_authenticated:
+            return redirect_to_login(request.get_full_path(), self.get_login_url(), self.get_redirect_field_name())
+        else:
+            if not self.is_precondition_satisfied():
+                return self.handle_no_permission()
+        return super().dispatch(request, *args, **kwargs)
+
+
+class ApplicationHubView(RedirectToLoginOrRelevantPageMixin, View):
+    """
+    View to display application hub, a checklist entailing different parts of the application process, and allow
+    users to submit their application.
+
+    """
+
+    template_name = 'adg/lms/applications/hub.html'
+    TOTAL_APPLICATION_OBJECTIVES = 2
+
+    def is_precondition_satisfied(self):
+        """
+        Checks if a user's application is already submitted or not.
+
+        Returns:
+            Boolean, True or False.
+
+        """
+        user_application_hub, _ = ApplicationHub.objects.get_or_create(user=self.request.user)
+
+        if self.request.method == 'POST':
+            return user_application_hub.is_prerequisite_courses_passed \
+                   and user_application_hub.is_written_application_completed
+        else:
+            return not user_application_hub.is_application_submitted
+
+    def handle_no_permission(self):
+        """
+        Redirects on test failure, `is_precondition_satisfied()` returns False.
+
+        Returns:
+            HttpResponse object.
+
+        """
+        if self.request.method == "POST":
+            return HttpResponse(status=400)
+        else:
+            return redirect('application_success')
+
+    def get(self, request):
+        """
+        Send the context data i.e user_application_hub, pre_req courses, and percentage to the template for rendering.
+
+        Returns:
+            HttpResponse object.
+
+        """
+        user_application_hub, _ = ApplicationHub.objects.get_or_create(user=self.request.user)
+        pre_req_course_ids = CourseMeta.open_pre_req_courses.all()
+        pre_req_courses = [CourseOverview.get_from_id(course_id) for course_id in pre_req_course_ids]
+
+        percentage_of_objectives_completed = '{0:.0%}'.format(
+            user_application_hub.number_of_objectives_completed / self.TOTAL_APPLICATION_OBJECTIVES)
+
+        return render(request, self.template_name, context={
+            'user_application_hub': user_application_hub,
+            'pre_req_courses': pre_req_courses,
+            'percentage': percentage_of_objectives_completed
+        })
+
+    def post(self, request):
+        """
+        Submit user application, send mandrill email according to the Application Confirmation Email format. In the
+        end, it redirects to the application success page.
+
+        Returns:
+            HttpResponse object.
+
+        """
+        user_application_hub = ApplicationHub.objects.get(user=request.user)
+        if not user_application_hub.is_application_submitted:
```

Comment:
Isn't this redirections part already handled in the mixin? shouldn't we assume the one request is inside this method that means application is not yet submitted?

Reply:
No it is necessary in this case. In case of an error, post request gives a http400 error. There, only this one redirection has to be done in the view.